### PR TITLE
1797 psydata logical support

### LINKED ISSUE #1797: [PSyData] Support for logical
At least NAN testing does not work with logical variables, which recent lfric version appear to be using. All other psydata need to be checked as well.

Note for myself: logical default type appear to be 4 (at least for gfortran)

### PR DESCRIPTION
**Pull request by hiker**
Fixes #1797 by adding support for Logical data types to PSyData libraries. In detail:
- `read-only` test: convert logical to 0/1 and use this as checksum.
- `nan-test`: tread logical as int and disable ieee function calls that are only defined for reals
- `profile`: no change, since the profiling lib does not provide any functions for variables
- `extraction`: support writing and reading logical values. Since (afaik) NetCDF does not support logical, the logicals are converted to int (and back)

Note that no psyclone code was changes, so no tests were added. The problems with logical variables happens only at compile time (since the generic interface cannot find an implementation for logicals). I have added a logical scalar to the LFRic extraction example. Again, afaik there are no arrays of logicals possible, so I could not add a test for handling of logical arrays.

### CODE REVIEW

File: lib/process.py [hunk 1]
```diff
@@ -105,6 +105,7 @@
 TYPE_DATA = {"real": ("Real", "real(kind=real32)", 32),
              "double": ("Double", "real(kind=real64)", 64),
              "int": ("Int", "integer(kind=int32)", 32),
+             "logical": ("Logical", "Logical(kind=4)", 4),
```
Comment (arporter) [inline lib/process.py:108]: Please update the copyright date.

**Review by arporter — CHANGES_REQUESTED** [review]
All good. All examples (and tests) OK with compilation.
Just a few files that need their copyright dates updating.

**Comment by arporter** [comment]
I think the dev. guide needs updating slightly - particularly https://psyclone-dev.readthedocs.io/en/latest/psy_data.html#jinja-support-in-the-base-class.

File: examples/lfric/eg17/full_example_extract/testkern_w0_kernel_mod.f90 [hunk 2]
```diff
@@ -46,9 +46,10 @@ module testkern_w0_kernel_mod
 
   type, public, extends(kernel_type) :: testkern_w0_kernel_type
      private
-     type(arg_type), dimension(2) :: meta_args =       &
+     type(arg_type), dimension(3) :: meta_args =       &
```
[the remark this answers is not in the document]
Comment (hiker) [reply examples/lfric/eg17/full_example_extract/testkern_w0_kernel_mod.f90:49]: Oops. Fixed.

File: lib/process.py (hunk 1 above)
Comment (hiker) [reply lib/process.py:108]: Fixed.

**Comment by hiker** [comment]
> I think the dev. guide needs updating slightly - particularly https://psyclone-dev.readthedocs.io/en/latest/psy_data.html#jinja-support-in-the-base-class.

Indeed, I totally missed that. Fixed now.

File: examples/lfric/eg17/full_example_extract/main.x90 [hunk 3]
```diff
@@ -41,7 +41,7 @@ program main
     use fs_continuity_mod,      only: W0, W1, W2, W2V, W2H
     use field_mod,              only: field_type
     use testkern_w0_kernel_mod, only: testkern_w0_kernel_type
-    use constants_mod,          only: r_def, i_def
+    use constants_mod,          only: r_def, i_def, l_def
```
[the remark this answers is not in the document]
Comment (hiker) [reply examples/lfric/eg17/full_example_extract/main.x90:44]: Fixed.

**Review by arporter — APPROVED** [review]
All requested changes have been made.
Dev. guide has been updated and builds and looks fine.
Will proceed to merge.

**Pull request merged**